ci/ray_ci/pipeline/gap_filling_scheduler.py

```diff
@@ -0,0 +1,66 @@
+import subprocess
+from datetime import datetime, timedelta
+from typing import List, Dict, Optional, Any
+
+from pybuildkite.buildkite import Buildkite
+
+
+BRANCH = "master"
+
+
+class GapFillingScheduler:
+    """
+    This buildkite pipeline scheduler is responsible for scheduling gap filling builds
+    when the latest build is failing.
+    """
+
+    def __init__(
+        self,
+        buildkite_organization: str,
+        buildkite_pipeline: str,
+        buildkite_access_token: str,
+        days_ago: int = 1,
+    ):
+        self.buildkite_organization = buildkite_organization
+        self.buildkite_pipeline = buildkite_pipeline
+        self.buildkite = Buildkite()
+        self.buildkite.set_access_token(buildkite_access_token)
+        self.days_ago = days_ago
+
+    def _get_latest_commit_for_build_state(self, build_state: str) -> Optional[str]:
+        latest_commits = self._get_latest_commits()
+        builds = sorted(
+            [
+                build
+                for build in self._get_builds()
+                if build["state"] == build_state and build["commit"] in latest_commits
+            ],
+            key=lambda build: latest_commits.index(build["commit"]),
+        )
+        if not builds:
+            return None
+
+        return builds[0]["commit"]
+
+    def _get_latest_commits(self) -> List[str]:
+        return (
+            subprocess.check_output(
```

Comment:
maybe send in a work dir here.

- bazel run does not run in the dir of the workspace. it can read the location though, so you either needs to send in the dir, or needs to change the work dir
- this would need the dir if we want to handle more than one repos in the same python program.
- if we use a real git repo for testing, then that dir needs to be send in / set up here too.

Reply:
ah yes, forgot about cwd